### ISSUE TITLE
Unable to update chekbox state

### ISSUE THREAD

**Issue by j33tmane** [jonataslaw/get#79]
Check box state not updating..
(i have initiated the controller already)
here is my Controller
```
import 'package:get/get.dart';

class SignUpController extends GetController {
  static SignUpController get to => Get.find();
  bool passwordVisible  = true;
  bool termsAccepted = true;
  void visibilityOnOff() {
    if(passwordVisible)
      passwordVisible=false;
    else
      passwordVisible=true;
    update(this);
  }

  void toggleTermsConditions(){
    if(termsAccepted)
      termsAccepted=false;
    else
      termsAccepted=true;
    update(this);
  }
}
```

check box implementation in UI
```
GetBuilder<SignUpController>(
              builder: (_)=> Checkbox(
                  value: _.termsAccepted,
                  checkColor: Colors.green,
                  activeColor: Colors.white,
                    onChanged: (value) {
                      SignUpController.to.toggleTermsConditions();
                    }
                ),
              ),
```

**Comment by jonataslaw** [comment]
Thank you for opening an issue with a reproduction example.
With Get you should use init: SignUpController (), only once, and on the first widget that uses the controller.
If there are 50 widgets in the tree using the same controller, use the one above.
I tried to reproduce the error, however, the checkbox is working normally:

```dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';

void main() {
  runApp(GetMaterialApp(
    home: First(),
  ));
}

class SignUpController extends GetController {
  static SignUpController get to => Get.find();
  bool passwordVisible = true;
  bool termsAccepted = true;
  void visibilityOnOff() {
    if (passwordVisible)
      passwordVisible = false;
    else
      passwordVisible = true;
    update(this);
  }

  void toggleTermsConditions() {
    if (termsAccepted)
      termsAccepted = false;
    else
      termsAccepted = true;
    update(this);
  }
}

class First extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: GetBuilder<SignUpController>(
          init: SignUpController(),
          builder: (_) => Checkbox(
              value: _.termsAccepted,
              checkColor: Colors.green,
              activeColor: Colors.white,
              onChanged: (value) {
                SignUpController.to.toggleTermsConditions();
              }),
        ),
      ),
    );
  }
}

```

You probably did not initialize your SignUpController, or initialized it after rendering the checkbox.

Can you reproduce the problem with my example?

**Comment by j33tmane** [comment]
Thanks for quick reply..
I tried your example now i came to know its working properly but i think i am confused with controller initialization here is my screen that i m trying to achieve
![img](https://user-images.githubusercontent.com/20676577/81428236-9f6ccf00-9179-11ea-89c0-743cf4ec7494.png)
..
My sign up page..

```
import 'package:flutter/material.dart';
import 'package:eliteboard/controllers/auth_signup_controller.dart';
import  'package:eliteboard/widgets/form.dart';
import 'package:flutter/services.dart';
import 'package:eliteboard/utilities/constants.dart';
import 'package:get/get.dart';
class SignUp extends StatelessWidget {

  final _loginFormKey = GlobalKey<FormState>();

  Widget _paddingBetweenInputs(){
    return  SizedBox(
      height: 20.0,
    );
  }

  Widget _buildSignupBtn(context) {
    return GestureDetector(
      onTap: () => Get.back(),
      child: RichText(
        text: TextSpan(
          children: [
            TextSpan(
              text: 'Already have an Account? ',
              style: Theme.of(context).textTheme.display1,
            ),
            TextSpan(
              text: 'Login',
              style: TextStyle(
                color: Colors.white,
                fontSize: 18.0,
                fontWeight: FontWeight.bold,
              ),
            ),
          ],
        ),
      ),
    );
  }



  //password Input field
  Widget _passwordInput(context){
    return GetBuilder<SignUpController>(
      init: SignUpController(),
      builder: (_) => FormInputField(
        iconData:Icons.lock,
        hintText: 'Enter yout password',
        inputType: TextInputType.text,
        secureText:_.passwordVisible,
        suffixIcon:IconButton(
          icon: Icon(
            !_.passwordVisible
                ? Icons.visibility
                : Icons.visibility_off,
            color: Theme.of(context).accentColor,
            size: 20.0,
          ),
          onPressed: () {
            SignUpController.to.visibilityOnOff();
          },
        ),
      ),
    );
  }

  // input field for name
  Widget _nameInput(context){
    return    FormInputField(
      iconData: Icons.account_box,
      hintText: 'Enter your full name',
      inputType: TextInputType.text,
      suffixIcon: SizedBox.shrink(),
    );
  }

  Widget _contactInput(context){
    return    FormInputField(
      iconData: Icons.phone_android,
      hintText: 'Enter your mobile',
      inputType: TextInputType.number,
      suffixIcon: SizedBox.shrink(),
    );
  }

  //input field for email
  Widget _emailInput(context){
    return    FormInputField(
      iconData: Icons.email,
      hintText: 'Enter your email',
      inputType: TextInputType.emailAddress,
      suffixIcon: SizedBox.shrink(),
    );
  }

  //Remember me chekbox button
  Widget _buildRememberMeCheckbox() {
    return Container(
      height: 20.0,
      child: Row(
//        crossAxisAlignment: CrossAxisAlignment.center,
//        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Theme(
            data: ThemeData(unselectedWidgetColor: Colors.white),
            child: GetBuilder<SignUpController>(
              init: SignUpController(),
              builder: (_) => Checkbox(
                  value: _.termsAccepted,
                  checkColor: Colors.green,
                  activeColor: Colors.white,
                  onChanged: (value) {
                    SignUpController.to.toggleTermsConditions();
                  }),
            ),
          ),
          Text(
            'I agree to terms & conditions',
            style: kLabelStyle,
          ),
        ],
      ),
    );
  }

  Widget _signUpForm(context){
    return Form(
        key: _loginFormKey,
        child: Column(
            children: <Widget>[
              _nameInput(context),
              _paddingBetweenInputs(),
              _emailInput(context),
              _paddingBetweenInputs(),
              _passwordInput(context),
              _paddingBetweenInputs(),
              _contactInput(context),
              _paddingBetweenInputs(),
              _buildRememberMeCheckbox()
            ]
        )
    );
  }
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: AnnotatedRegion<SystemUiOverlayStyle>(
        value: SystemUiOverlayStyle.light,
        child: GestureDetector(
          onTap: () => FocusScope.of(context).unfocus(),
          child: Stack(
            children: <Widget>[
              AuthScreenFormBackground(),
              Container(
                height: double.infinity,
                child: SingleChildScrollView(
                  physics: AlwaysScrollableScrollPhysics(),
                  padding: EdgeInsets.symmetric(
                    horizontal: 40.0,
                    vertical: 80.0,
                  ),
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      FormPageTitle(
                          title:'ELITEBOARD'
                      ),
                      SizedBox(height: 40.0),
                      _signUpForm(context),
                      FormSubmitButton(
                        onPressed: (){
                          if (_loginFormKey.currentState.validate()) {

                          }
                        },
                        buttonTitle: 'REGISTER',
                      ),
                      _buildSignupBtn(context),
                    ],
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
```

My widgets/form.dart file

```
import 'package:flutter/material.dart';
import 'package:eliteboard/utilities/constants.dart';


class FormInputField extends StatelessWidget {
  final IconData iconData;
  final String hintText;
  final TextInputType inputType;
  final bool secureText;
  final Widget suffixIcon;
  FormInputField({@required this.iconData, @required this.hintText,@required this.inputType,this.secureText=false,@required this.suffixIcon});
  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        Container(
          alignment: Alignment.centerLeft,
          decoration: kBoxDecorationStyle,
          height: 60.0,
          child: TextFormField(
            autovalidate: false,
            validator: (value) {
              if (value.isEmpty) {
                return ' Input field required!';
              }
              return null;
            },
            keyboardType: inputType,
            obscureText: secureText,
            style: TextStyle(
              color: Colors.white,
              fontFamily: 'OpenSans',
            ),
            decoration: InputDecoration(
              errorStyle: TextStyle(
                color: Theme.of(context).errorColor,
                wordSpacing: 5.0,

              ),
              border: InputBorder.none,
              contentPadding: EdgeInsets.only(top: 14.0),
              prefixIcon: Icon(
                iconData,
                color: Colors.white,
              ),
              suffixIcon: suffixIcon,
              hintText: hintText,
              hintStyle: kHintTextStyle,

            ),
          ),
        ),
      ],
    );
  }
}

class FormSubmitButton extends StatelessWidget{
  final String buttonTitle;
  final VoidCallback onPressed;
  FormSubmitButton({@required this.buttonTitle, @required this.onPressed});
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(vertical: 25.0),
      width: double.infinity,
      child: RaisedButton(
        elevation: 5.0,
        onPressed:onPressed,
        padding: EdgeInsets.all(15.0),
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(30.0),
        ),
        color: Colors.white,
        child: Text(
          buttonTitle,
          style: TextStyle(
            color: Color(0xFF527DAA),
            letterSpacing: 1.5,
            fontSize: 18.0,
            fontWeight: FontWeight.bold,
            fontFamily: 'OpenSans',
          ),
        ),
      ),
    );
  }
}

class AuthScreenFormBackground extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
   return Container(
      height: double.infinity,
      width: double.infinity,
      decoration: BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: [
            Color(0xFF73AEF5),
            Color(0xFF61A4F1),
            Color(0xFF478DE0),
            Color(0xFF398AE5),
          ],
          stops: [0.1, 0.4, 0.7, 0.9],
        ),
      ),
    );
  }
}

class FormPageTitle extends StatelessWidget{
  final String title;
  FormPageTitle({@required this.title});
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.only(top: 20.0),
      child: Text(
        title,
        style: TextStyle(
          color: Colors.white,
          fontFamily: 'OpenSans',
          fontSize: 30.0,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}
```

Constants file 

```
import 'package:flutter/material.dart';

final kHintTextStyle = TextStyle(
  color: Colors.white54,
  fontFamily: 'OpenSans',
);

final kLabelStyle = TextStyle(
  color: Colors.white,
  fontWeight: FontWeight.bold,
  fontFamily: 'OpenSans',
);

final kBoxDecorationStyle = BoxDecoration(
  color: Color(0xFF6CA8F1),
  borderRadius: BorderRadius.circular(10.0),
  boxShadow: [
    BoxShadow(
      color: Colors.black12,
      blurRadius: 6.0,
      offset: Offset(0, 2),
    ),
  ],
);
```

**Comment by jonataslaw** [comment]
I try you code and work to me:
https://www.loom.com/share/32b39b56391148dcb2443c9670b54f85

Are you using the latest version of Flutter and the latest version of Get?

**Comment by jonataslaw** [comment]
> I try you code and work to me:
> 
> https://www.loom.com/share/32b39b56391148dcb2443c9670b54f85
> 
> Are you using the latest version of Flutter and the latest version of Get?

And just a tip (not related to issue)
Putting the formkey inside your controller is a matter of good practice. The way you are using it works perfectly, but your key is inside a build method, and will not be discarded when the screen comes out of the widget tree (it is an insignificant memory leak for your example, but if you continue to do so at the project will not be a good thing).
I know it can be boring to call the controller for everything, but by doing this, you will program with 100% certainty that you will not have to worry about optimizing your app in the future.

It is also not considered good practice to use functional widgets (` Widget _antThing () {return Foo ();}`), because with each class reconstruction they are created again and in a stateful that stores state, the state of the widgets created previously goes to Narnia and can never be discarded again, but this only affects the conventional Flutter because it has Stateful classes, as Get is completely StatelessWidget, and only GetBuilder is a stateful, this will not be a problem with Get (if someone complains about you in the future , you can use "I use Get" as an argument, but remember that without Get, this can cause memoryleakes)

**Comment by j33tmane** [comment]
I update everything now its working but My old router was giving error 
i.e. A value of type 'GetRoute' can't be assigned to a variable of type 'Route<dynamic>'. 
So i shifted to namedRoutes in material app insted of onGenerateRoute : Router.generateRoute()

**Comment by j33tmane** [comment]
> 
> 
> > I try you code and work to me:
> > https://www.loom.com/share/32b39b56391148dcb2443c9670b54f85
> > Are you using the latest version of Flutter and the latest version of Get?
> 
> And just a tip (not related to issue)
> Putting the formkey inside your controller is a matter of good practice. The way you are using it works perfectly, but your key is inside a build method, and will not be discarded when the screen comes out of the widget tree (it is an insignificant memory leak for your example, but if you continue to do so at the project will not be a good thing).
> I know it can be boring to call the controller for everything, but by doing this, you will program with 100% certainty that you will not have to worry about optimizing your app in the future.
> 
> It is also not considered good practice to use functional widgets (` Widget _antThing () {return Foo ();}`), because with each class reconstruction they are created again and in a stateful that stores state, the state of the widgets created previously goes to Narnia and can never be discarded again, but this only affects the conventional Flutter because it has Stateful classes, as Get is completely StatelessWidget, and only GetBuilder is a stateful, this will not be a problem with Get (if someone complains about you in the future , you can use "I use Get" as an argument, but remember that without Get, this can cause memoryleakes)

great i wll try to update my coding standards, its nice of you to guide a noob like me.. Thanks bro.

**Comment by jonataslaw** [comment]
> I update everything now its working but My old router was giving error
> i.e. A value of type 'GetRoute' can't be assigned to a variable of type 'Route'.
> So i shifted to namedRoutes in material app insted of onGenerateRoute : Router.generateRoute()

Flutter's onGenerateRoute is completely broken and buggy in the dev / master branch, so I created this namedRoutes wrap to handle the routes and modify them, I recommend using it, because even if you are in the Flutter's master/dev branch, it will work , in addition to making the code cleaner, as it is just a map. with name and route.

Is the checkbox working now? Can I close the issue?

**Comment by jonataslaw** [comment]
Based on your comment that the update solved for you and the reproduction of the problem that worked as expected with me, I am closing this issue.
If you face any problems with Get, do not hesitate to open another issue. If there are still doubts about something related to the issue, you can comment here that I will answer.
